src/proto/streams/recv.rs

```diff
@@ -368,19 +381,33 @@ impl Recv {
 
         // Ensure that there is enough capacity on the connection before acting
         // on the stream.
-        if self.flow.window_size() < sz || stream.recv_flow.window_size() < sz {
-            return Err(RecvError::Connection(Reason::FLOW_CONTROL_ERROR));
+        self.consume_connection_window(sz)?;
+
+        if is_ignoring_frame {
+            trace!("recv_data frame being ignored on locally reset {:?} for some time", stream.id);
+            return Ok(());
         }
 
-        // Update connection level flow control
-        self.flow.send_data(sz);
+        if stream.recv_flow.window_size() < sz {
+            // http://httpwg.org/specs/rfc7540.html#WINDOW_UPDATE
+            // > A receiver MAY respond with a stream error (Section 5.4.2) or
+            // > connection error (Section 5.4.1) of type FLOW_CONTROL_ERROR if
+            // > it is unable to accept a frame.
+            //
+            // So, for violating the **stream** window, we can send either a
+            // stream or connection error. We've opted to send a stream
+            // error.
+            return Err(RecvError::Stream {
```

Comment:
Similar as above, we need to be very sure that if this code path is hit, the connection level window is not leaked. Again, it would be nice to have a test for this.

[Related](https://github.com/carllerche/h2/pull/174/files#r151241765)

Reply:
Looking around this function, it appears there's existing conditions that could cause a stream error after consuming the connection window, and in all of those cases, the window is "leaked".

To fix this, we could add a check outside this function, wherever `streams.recv_data()` is being called, and if a stream error is returned, try to release the connection capacity again...